### PR TITLE
Handle invalid date of birth for participant validation

### PR DESCRIPTION
**Pull request by xswoodsx**
## Ticket and context

Ticket: [882](https://dfedigital.atlassian.net/jira/software/projects/CPDEL/boards/102?assignee=5f3543bb3e9e2e004db3189d&selectedIssue=CPDRP-882)

### Is there anything that the code reviewer should know?

Dates aren't fun.

`ActiveRecord::Type::Date.new.cast(value)` will, for example, go to 1st October 1991 when putting 31 September but Date.new(31, 9, 1991) will raise an invalid error. 

Happy to chat about if this if anyone if free. I have a couple questions about the form and how it's handling the dob.

### Code quality checks
- [ ] All commit messages are meaningful and true
- [ ] Added enough automated tests

### HTML Checks
- [ ] All new pages have automated accessibility checks
- [ ] All new pages have visual tests via Percy

### Gotchas
- [ ] All `School` queries are correctly scoped - `eligible` when they need to be

### How can someone see it it review app?
1. Click the link to review app (posted by a `github-actions` bot below)
2. login as an example email 
3. go to `/participants/validation/tell-us-your-details`
4. Try various different d.o.b's, valid and invalid

## External API changes

Does this change anything in our external APIs? If so, did you remember to update the CHANGELOG for Lead Providers? (lead-providers/guidance/release-notes)

### CODE REVIEW

File: app/forms/participants/participant_validation_form.rb [hunk 1]
```diff
@@ -45,6 +45,10 @@ def attributes
     def date_of_birth=(value)
       @date_of_birth_invalid = false
       @date_of_birth = ActiveRecord::Type::Date.new.cast(value)
+
+      if invalid_date?(value)
+        @date_of_birth_invalid = true
+      end
```
Comment (ethanmills) [inline app/forms/participants/participant_validation_form.rb:51]: Do we want to move this before the assignment and add an early return?

Comment (xswoodsx) [reply app/forms/participants/participant_validation_form.rb:51]: Returning early is causing problems with some tests, notably feature tests... doing `@date_of_birth = value` in the check, makes it blank but with the invalid date message.

File: app/forms/participants/participant_validation_form.rb [hunk 2]
```diff
@@ -45,6 +45,11 @@ def attributes
     def date_of_birth=(value)
       @date_of_birth_invalid = false
       @date_of_birth = ActiveRecord::Type::Date.new.cast(value)
+
+      if invalid_date?(value)
+        @date_of_birth_invalid = true
+        @date_of_birth = value
+      end
```
Comment (jamesrose) [inline app/forms/participants/participant_validation_form.rb:52]: I don't think you need `invalid_date?` here. I'm pretty sure that `ActiveRecord::Type::Date.new.cast` will return `nil` if the date is not valid.

Comment (xswoodsx) [reply app/forms/participants/participant_validation_form.rb:52]: If you enter 31/9/1990, for example, it amends the date to 01/10/1990. `Date.new(1990/9/31)` does return nil. Not sure what ActiveRecord is doing differently but something about 1 day over how many is in that month puts it to the 1st of the next month.

**Review by ethanmills — APPROVED** [review]

**Pull request merged**